File: web-service/src/main/java/io/acari/event/source/handler/PodHandler.java
package io.acari.event.source.handler;


import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.acari.event.source.models.*;
import io.acari.event.source.repository.PodMemberRepository;
import io.acari.event.source.repository.PodRepository;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class PodHandler {
    private final PodRepository podRepository;
    private final PodMemberRepository podMemberRepository;
    private final ObjectMapper objectMapper;

    public PodHandler(PodRepository podRepository, PodMemberRepository podMemberRepository, ObjectMapper objectMapper) {
        this.podRepository = podRepository;
        this.podMemberRepository = podMemberRepository;
        this.objectMapper = objectMapper;
    }

    public Stream<Identifier> projectAllPodMembers() {
        return podRepository.allPodEvents()
                .collect(HashSet<String>::new,
                        (podMembers, podEvent) -> {
                            try {
                                BasePodMemberPayload basePodMemberPayload = objectMapper.treeToValue(podEvent.getPayload(), BasePodMemberPayload.class);
                                if (EventTypes.POD_MEMBER_DELETED.equals(podEvent.getType())) {
                                    podMembers.remove(basePodMemberPayload.getIdentifier());
                                } else {
                                    podMembers.add(basePodMemberPayload.getIdentifier());
                                }
                            } catch (JsonProcessingException e) {
                                e.printStackTrace();
                            }
                        }, Set::addAll).stream().map(Identifier::new);
    }

    public Optional<PersonalInformation> projectPersonalInformation(String podMemberIdentifier) {
        List<Event> podMemberEvents = podMemberRepository.fetchPodMemberEventStream(podMemberIdentifier).collect(Collectors.toList());
        PersonalInformation projectedPersonalInformation = podMemberEvents.stream()
                .filter(event -> EventTypes.PERSONAL_INFO_CAPTURED.equals(event.getType()))
                .map(Event::getPayload)
                .flatMap(payload -> {
                    try {
                        return Stream.of(objectMapper.treeToValue(payload, CapturedInfoPayload.class));
                    } catch (JsonProcessingException e) {
                        return Stream.empty();
                    }
                })
                .collect(PersonalInformation::new, (personalInformation, capturedInfoPayload) -> {
                    switch (capturedInfoPayload.getField()) {
                        case "email":
                            personalInformation.setEmail(capturedInfoPayload.getValue());
                            break;
                        case "firstName":
                            personalInformation.setFirstName(capturedInfoPayload.getValue());
                            break;
                        case "lastName":
                            personalInformation.setLastName(capturedInfoPayload.getValue());
                            break;
                        case "phoneNumber":
                            personalInformation.setPhoneNumber(capturedInfoPayload.getValue());
                            break;
                    }
                }, (personalInformation, personalInformation2) -> {
                });

        return Optional.of(projectedPersonalInformation);
    }

    public Optional<Event> savePodMemberEvent(String podMemberIdentifier, Event eventToSave) {
        return podMemberRepository.saveEvent(podMemberIdentifier, eventToSave);
    }

    /**
     * Accepts a String which is literally just JSON.
     * I ran into an issue of not being able to create a JSONNode from
     * the REST endpoint. ¯\_(ツ)_/¯
     * <p>
     * But this works!
     */
    public Optional<String> savePodEvent(String eventAsJson) {
        return podRepository.saveEvent(eventAsJson);
    }


}

File: web-service/src/main/kotlin/io/acari/event/source/handler/PodHandler.kt
package io.acari.event.source.handler

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.acari.event.source.models.Event
import io.acari.event.source.models.Identifier
import io.acari.event.source.models.PersonalInformation
import io.acari.springwebflux.models.*
import io.acari.event.source.repository.PodMemberRepository
import io.acari.event.source.repository.PodRepository
import org.reactivestreams.Publisher
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.toMono
import java.util.*

/**
 * Forged in the flames of battle by alex.
 */
@Service
class PodHandler(
    private val podRepository: PodRepository,
    private val podMemberRepository: PodMemberRepository,
    private val imageHandler: ImageHandler) {

  private val objectMapper = jacksonObjectMapper()
      .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

  fun projectAllPodMembers(): Flux<Identifier> =
      Flux.empty()

  fun projectPersonalInformation(podMemberIdentifier: String): Mono<PersonalInformation> {
    return PersonalInformation().toMono()
  }

  fun fetchAvatar(podMemberIdentifier: String): Flux<ByteArray> =
      Flux.empty()

  fun savePodMemberEvent(podMemberIdentifier: String, bodyToMono: Mono<Event>): Publisher<Event> =
      bodyToMono

  /**
   * Accepts a String which is literally just JSON.
   * I ran into an issue of not being able to create a JSONNode from
   * the functional endpoint. ¯\_(ツ)_/¯
   *
   * But this works!
   */
  fun savePodEvent(bodyToMono: Mono<String>): Publisher<String> =
      bodyToMono
}